### ISSUE TITLE
Screen reader can't stay focused on languages menu.

### ISSUE THREAD

**Issue by rkingett** [liberapay/liberapay.com#1774]
Using [NVDA,](https://www.nvaccess.org/about-nvda) an open source screen reader, on [this profile,](https://en.liberapay.com/Liberapay/) for example, NVDA treats the languages menu as one giant level 5 heading, but it is very difficult for the screen reader to stay focused in that languages menu.

I can tell there's a button. In my case, the English button. I'm guessing that's the language currently selected? I wasn't quite sure.

When trying to interact with the language menu on any profile, the screen reader loses focus and jumps around the page wildly.

Perhaps, making the language selections a simple HTML 5 combo box would be the best move.

**Comment by Changaco** [comment]
I don't have any experience with screen readers, but using keyboard navigation I can reach the language links inside the dropdown menu by pressing the down-arrow key twice when the focus is on the button labeled “English” (which is indeed the selected language).

It's possible that dropdowns are more accessible in newer versions of Bootstrap (#940).

Using an HTML `<select>` would be less friendly when JavaScript is disabled.

**Comment by rkingett** [comment]
I had to switch NVDA to [focus mode](https://tink.uk/understanding-screen-reader-interaction-modes/) to tab all the way to the languages menu, the minute I switch back to Browse mode, it causes focus to jump all over the page.

Narrator, a built in screen reader for windows 10, seems to completely ignore the menu when trying to expand that menu in Microsoft Edge and Chrome.

I hope bootstrap becomes more accessible by way of drop down menus. I can attach a video if that would help?

**Comment by Changaco** [comment]
Here's what the documentation for the latest version of Bootstrap says about the accessibility of dropdowns:

> The [WAI ARIA](https://www.w3.org/TR/wai-aria/) standard defines an actual [`role="menu"` widget](https://www.w3.org/WAI/PF/aria/roles#menu), but this is specific to application-like menus which trigger actions or functions. ARIA menus can only contain menu items, checkbox menu items, radio button menu items, radio button groups, and sub-menus.
> 
> Bootstrap’s dropdowns, on the other hand, are designed to be generic and applicable to a variety of situations and markup structures. For instance, it is possible to create dropdowns that contain additional inputs and form controls, such as search fields or login forms. For this reason, Bootstrap does not expect (nor automatically add) any of the `role` and `aria-` attributes required for true ARIA menus. Authors will have to include these more specific attributes themselves.
> 
> However, Bootstrap does add built-in support for most standard keyboard menu interactions, such as the ability to move through individual `.dropdown-item` elements using the cursor keys and close the menu with the ESC key.

https://getbootstrap.com/docs/4.5/components/dropdowns/#accessibility

Looking at the examples and comparing them to our code, I see that we're missing the `aria-haspopup="true"` and `aria-expanded="false"` attributes on the button, and that we have an extra `role="menu"` attribute on the list.

**Comment by Changaco** [comment]
@rkingett Do the dropdowns in the top navigation bar also confuse the screen reader? They're built the same way as the language menu of a profile page, but they're in a different section of the page (in a `<nav>` tag).

**Comment by rkingett** [comment]
Neither Narrator nor NVDA get tripped up when reviewing that top level menu, that navigation menu, but I'm guessing that actually has no ARIA at all? Which would be ideal, actually. Symantic, native, code, should work just fine, maybe?

That button at the top is not labeled though. There's an unlabeled graphic at the top too. I wrote about them in separate issues.

**Comment by Jookia** [comment]
@Changaco I highly suggest installing NVDA or Orca and testing the site yourself.

Also, role=menu is not the correct solution for this. See https://adrianroselli.com/2017/10/dont-use-aria-menu-roles-for-site-nav.html and https://adrianroselli.com/2020/03/stop-using-drop-down.html

**Comment by rkingett** [comment]
Agreed. I also wrote about that unlabeled check box in another issue, as well as that weird top menu. Do not change native semantics, unless you really have to.
